rattan-core/src/cells/delay.rs

```diff
@@ -296,60 +281,59 @@ where
         #[cfg(not(feature = "first-packet"))]
         crate::wait_until_started!(self, Start);
 
-        let packet = loop {
-            tokio::select! {
-                biased;
-                Some(config) = self.config_rx.recv() => {
-                    self.set_config(config);
-                }
-                _ = self.change_timer.sleep(self.next_change - Instant::now()) => {
-                    if self.next_change(self.next_change).is_none() {
-                        debug!("Trace goes to end");
-                        self.egress.close();
-                        return None;
-                    }
-                }
-                packet = self.egress.recv() => if let Some(packet) = packet { break packet }
-            }
-        };
-        match self.state.load(std::sync::atomic::Ordering::Acquire) {
-            0 => {
-                return None;
-            }
-            1 => {
-                return Some(packet);
-            }
-            _ => {}
+        // It could be None only if the other end of the channel has closed.
+        let packet = self.egress.recv().await?;
+        while let Ok(config) = self.config_rx.try_recv() {
+            self.set_config(config);
         }
-        self.next_available = packet.get_timestamp() + self.current_delay;
-        loop {
-            tokio::select! {
-                biased;
-                Some(config) = self.config_rx.recv() => {
-                    self.set_config(config);
-                }
-                _ = self.change_timer.sleep(self.next_change - Instant::now()) => {
-                    if self.next_change(self.next_change).is_none() {
-                        debug!("Trace goes to end");
-                        self.egress.close();
-                        return None;
-                    }
-                }
-                _ = self.send_timer.sleep(self.next_available - Instant::now()) => {
-                    break;
-                }
-            }
+
+        let mut packet = crate::check_cell_state!(self.state, packet);
+
+        let timestamp = packet.get_timestamp();
+
+        // Update the config until it is applicable for the packet.
+        while self.next_change <= timestamp {
+            self.update_delay(self.next_change);
         }
+
+        let send_time = {
+            // `get_current` returns None if and only if `self.delays` has never been updated
+            // since last reset.
+
+            let logical_send_time = if let Some(delay_time) = self.delays.get_current(timestamp) {
+                (timestamp + *delay_time).max(self.latest_egress_timestamp)
+            } else {
+                timestamp.max(self.latest_egress_timestamp)
+            };
+
+            let sleep_time = logical_send_time.duration_since(Instant::now());
+
+            if !sleep_time.is_zero() {
+                let _ = self.send_timer.sleep(sleep_time).await;
+            }
```

Comment:
It isn't correct that you remove the change_timer: what if the delay trace changes during your sleep? This is what I have discussed with you: your local testing might not extensively explore all cases, so you should logically construct all possible cases to handle.